### ISSUE TITLE
Documentation is missing information on ABI stability

### ISSUE THREAD

**Issue by christianglodt** [tree-sitter/tree-sitter#374]
I'm missing some information regarding ABI stability that I think it would be helpful to add to the docs.

I'm creating an Atom language package which depends on an npm package with a tree-sitter based parser. I'm unsure which tree-sitter version to use in order to make my grammar package work in as many Atom versions as possible.

Atom 1.37.0 uses the tree-sitter runtime 0.13.23, while Atom 1.38.2 uses tree-sitter runtime 0.15.4.

If I prebuild with 0.13.23, my grammar will crash Atom 1.38.2 (at least on MacOS). If I prebuild with 0.15.7, my grammar will not be recognized by Atom 1.37.0 (ie. I get no syntax highlighting, incompatible language version message in console).

Would it be possible to document how ABI stability is handled in tree-sitter?

Am I right to assume that LANGUAGE_VERSION defines an ABI version?

Which versions support which LANGUAGE_VERSIONs? Also, which Atom versions would support grammars created with which tree-sitter version? Is there a sweet spot that would allow me to support most Atom versions?

Thanks for clarifying!

**Comment by maxbrunsfeld** [comment]
I agree this is something that definitely should be added to the docs. I'll try to write it up when I get a chance.

> Am I right to assume that LANGUAGE_VERSION defines an ABI version?

Yes.

> Which versions support which LANGUAGE_VERSIONs?

The Tree-sitter library can be used with language parsers that were built with an *older* version of Tree-sitter, but the reverse does not work. The exact language versions that work with the current library version are defined by [these two constants](https://github.com/tree-sitter/tree-sitter/blob/34de25ce54c99b76d9ad4628af5c2569dd737b6b/lib/include/tree_sitter/api.h#L17-L18) in `api.h`.

Does that make sense?

**Comment by christianglodt** [comment]
Thanks, yes that does make sense.

**Comment by maxbrunsfeld** [comment]
I've added some docs in the header file:

https://github.com/tree-sitter/tree-sitter/blob/a1c3de1b9976cd20ebd158d05ad630f794df2490/lib/include/tree_sitter/api.h#L17-L30.

I think there should probably still be some mention of these things on the docs site.

**Comment by Thaodan** [comment]
I think this isn't just a documentation issue but also an issue of the libtreesittter library not tracking the abi of grammars it support and the generated grammar not containing the abi version they are built against.
There should be a direct relation between them.

**Comment by ObserverOfTime** [comment]
> the libtreesittter library not tracking the abi of grammars it support and the generated grammar not containing the abi version they are built against.

Neither of those is true.

**Comment by Thaodan** [comment]
Is false now or at the time?

**Comment by wetneb** [comment]
I made a table summarizing the ABI versions supported by recent versions of tree-sitter when loading a parser:
| tree-sitter version | Min parser ABI version | Max parser ABI version |
|---------------------|------------------------|------------------------|
| 0.14                | 9                      | 9                      |
| >=0.15.0, <=0.15.7  | 9                      | 10                     |
| >=0.15.8, <=0.16    | 9                      | 11                     |
| 0.17, 0.18          | 9                      | 12                     |
| >=0.19, <=0.20.2    | 13                     | 13                     |
| >=0.20.3, <=0.24    | 13                     | 14                     |
| >=0.25              | 13                     | 15                     |

(I hope it's correct, it can be that I made mistakes…)

Would it be helpful if I made a PR to add a page to the docs with this information? I imagine it as a page in the `Using parsers` section, towards the end.

Ideally, it would be nice if we could give users a sense of what to expect concerning forthcoming ABI bumps. I don't have a crystal ball and I don't expect the project to make firm commitments about how long a particular ABI version would be supported for instance, but we could try to answer the following questions:
* are there any planned features that are likely to require a new ABI? (with a link to a GitHub issue). Even without a planned release date or corresponding tree-sitter version number, I think it would already be interesting to have that overview.
* would the existing project team find it okay to publish a new version of tree-sitter with a new ABI and no compatibility with the previous ABIs, similarly to 0.19?

**Comment by sogaiu** [comment]
FWIW, it's not as complete as the table above, but there is some partial similar info [in a table here](https://github.com/sogaiu/ts-questions/tree/43297a480c8de1a7ba64cdd48ff2ba7f132ecd4a/questions/what-abi-level-should-be-used#discussion).  The purpose of this latter table is a bit different but there is some overlap in content (but also there are date bits).

The info may also be off in parts 😅
